dcrlibwallet.go

```diff
@@ -676,63 +711,149 @@ func (lw *LibWallet) GetBestBlockTimeStamp() int64 {
+		}
+
+		select {
+		case <-ctx.Done():
+			return true, ctx.Err()
+		default:
+			return false, nil
+		}
+	}
+
+	if beginHeight == -1 {
+		var previousEndBlock int32
+		err := lw.db.Get(BucketTxInfo, KeyEndBlock, &previousEndBlock)
+		if err != nil && err != storm.ErrNotFound {
+			log.Errorf("Get not found :%v", err)
+			return err
+		}
+
+		if previousEndBlock != 0 {
+			beginHeight = previousEndBlock
+			beginHeight -= MaxReOrgBlocks
+
+			if beginHeight < 0 {
+				beginHeight = 0
+			}
+		} else {
+			beginHeight = 0
+		}
+	}
+
+	if beginHeight > endHeight {
+		endHeight = lw.GetBestBlock()
+	}
+
+	startBlock := wallet.NewBlockIdentifierFromHeight(beginHeight)
+	endBlock := wallet.NewBlockIdentifierFromHeight(endHeight)
+
+	defer func() {
+		for _, syncResponse := range lw.syncResponses {
+			syncResponse.OnSynced(true)
+		}
+		count, err := lw.db.Count(&Transaction{})
+		if err != nil {
+			log.Errorf("Count Error :%v", err)
+			return
+		}
+		log.Infof("Transaction index finished at %d, %d transaction(s) indexed in total", endHeight, count)
+	}()
+
+	log.Infof("Indexing transactions start height: %d, end height: %d", beginHeight, endHeight)
+	return lw.wallet.GetTransactions(rangeFn, startBlock, endBlock)
+}
+
 func (lw *LibWallet) TransactionNotification(listener TransactionListener) {
 	go func() {
 		n := lw.wallet.NtfnServer.TransactionNotifications()
 		defer n.Done()
 		for {
 			v := <-n.C
 			for _, transaction := range v.UnminedTransactions {
-				var amount int64
-				var inputAmounts int64
-				var outputAmounts int64
-				tempCredits := make([]TransactionCredit, len(transaction.MyOutputs))
-				for index, credit := range transaction.MyOutputs {
-					outputAmounts += int64(credit.Amount)
-					tempCredits[index] = TransactionCredit{
-						Index:    int32(credit.Index),
-						Account:  int32(credit.Account),
-						Internal: credit.Internal,
-						Amount:   int64(credit.Amount),
-						Address:  credit.Address.String()}
-				}
-				tempDebits := make([]TransactionDebit, len(transaction.MyInputs))
-				for index, debit := range transaction.MyInputs {
-					inputAmounts += int64(debit.PreviousAmount)
-					tempDebits[index] = TransactionDebit{
-						Index:           int32(debit.Index),
-						PreviousAccount: int32(debit.PreviousAccount),
-						PreviousAmount:  int64(debit.PreviousAmount),
-						AccountName:     lw.AccountName(int32(debit.PreviousAccount))}
+
+				tempTransaction, err := lw.parseTxSummary(&transaction, nil)
+				if err != nil {
+					log.Errorf("Error ntfn parse tx: %v", err)
+					return
 				}
-				var direction int32
-				amountDifference := outputAmounts - inputAmounts
-				if amountDifference < 0 && (float64(transaction.Fee) == math.Abs(float64(amountDifference))) {
-					//Transfered
-					direction = 2
-					amount = int64(transaction.Fee)
-				} else if amountDifference > 0 {
-					//Received
-					direction = 1
-					amount = outputAmounts
+
+				var oldTx Transaction
```

Comment:
Lines 836 to 854 is repeated multiple times and can be refactored into a function. It also appears in lines 734 - 753 and 875 to 894